File: Exercism/haskell/acronym/src/Acronym.hs
module Acronym (abbreviate) where

{-# LANGUAGE OverloadedStrings #-}
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Char as C


abbreviate :: Text -> Text
abbreviate = T.toUpper . T.filter C.isAlpha . snd . T.mapAccumL dispatch ' '
  where dispatch prev c
          | not $ C.isAlpha c = (c, ' ')
          | prev /= '\'' && (C.isSpace prev || C.isPunctuation prev) = (c, c)
          | C.isLower prev && C.isUpper c = (c, c)
          | otherwise = (c, ' ')

File: Exercism/c/acronym/src/acronym.c
#include "acronym.h"
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#include <stdio.h>


char *abbreviate(const char *phrase) {
  if (phrase == NULL)
    return NULL;

  char * acronym = malloc(strlen(phrase) + 1);
  char *ap = acronym;
  char prev = ' ';

  for(char *c = (char *)phrase; *c; prev = *c, c++) {
    if (isalpha(*c) && prev != '\'' && (isspace(prev) || ispunct(prev)))
     *ap++ = (unsigned char)toupper(*c);
  }
  if (ap == acronym) {
    free(acronym);
    return NULL;
  }
  *ap = '\0';
  return acronym;
}
